File: scripts/build_pack.py
#!/usr/bin/env python3
"""Build a 1.16.1 language resource pack from Crowdin output + pinned en_us.json."""

from __future__ import annotations

import argparse
import json
import re
import sys
import zipfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SOURCE_EN = ROOT / "source" / "en_us.json"
PACK_MCMETA = ROOT / "pack" / "pack.mcmeta"
TRANSLATIONS_DIR = ROOT / "translations"
DIST_DIR = ROOT / "dist"
DIST_FONT = DIST_DIR / "linzi_font"
ZIP_NAME = "pekzep-1.16.1.zip"
LINZI_ZIP_NAME = "pekzep-linzi-1.16.1.zip"
LINZI_DESCRIPTION = "牌言・燐字 (Pekzep linzi) for Minecraft 1.16.1"

PLACEHOLDER = re.compile(r"%(?:\d+\$)?[sd]|%%")


def load_json(path: Path) -> dict:
    with path.open(encoding="utf-8") as f:
        data = json.load(f)
    if not isinstance(data, dict):
        raise SystemExit(f"{path} is not a JSON object")
    return data


def find_translation_file(explicit: Path | None) -> Path:
    if explicit is not None:
        if not explicit.is_file():
            raise SystemExit(f"translation file not found: {explicit}")
        return explicit

    preferred = [
        TRANSLATIONS_DIR / "pz_ai.json",
        TRANSLATIONS_DIR / "pz_ai" / "en_us.json",
        TRANSLATIONS_DIR / "pz-ai" / "en_us.json",
        TRANSLATIONS_DIR / "pz-AI" / "en_us.json",
    ]
    for path in preferred:
        if path.is_file():
            return path

    if TRANSLATIONS_DIR.is_dir():
        candidates = sorted(
            p for p in TRANSLATIONS_DIR.rglob("*.json") if p.is_file()
        )
        if len(candidates) == 1:
            return candidates[0]
        if candidates:
            pz = [
                p
                for p in candidates
                if "pz" in str(p.relative_to(TRANSLATIONS_DIR)).lower()
            ]
            if len(pz) == 1:
                return pz[0]
            names = ", ".join(str(p.relative_to(ROOT)) for p in candidates)
            raise SystemExit(
                "multiple translation JSON files; pass --translations. found: "
                + names
            )

    raise SystemExit(
        "no translation file. download from Crowdin into translations/"
    )


def language_meta(mcmeta: dict) -> tuple[str, str, str]:
    langs = mcmeta.get("language")
    if not isinstance(langs, dict) or len(langs) != 1:
        raise SystemExit("pack.mcmeta must define exactly one language")
    code, info = next(iter(langs.items()))
    if not isinstance(info, dict):
        raise SystemExit(f"invalid language entry for {code}")
    name = info.get("name")
    region = info.get("region")
    if not name or not region:
        raise SystemExit("language name and region are required in pack.mcmeta")
    return code, name, region


def merge(source: dict, translations: dict) -> dict:
    merged = dict(source)
    for key, value in translations.items():
        if isinstance(value, str) and value != "":
            merged[key] = value
    return merged


def warn_placeholders(source: dict, merged: dict) -> int:
    warnings = 0
    for key, src in source.items():
        if not isinstance(src, str):
            continue
        dst = merged.get(key)
        if not isinstance(dst, str) or dst == src:
            continue
        src_ph = PLACEHOLDER.findall(src)
        dst_ph = PLACEHOLDER.findall(dst)
        if src_ph != dst_ph:
            warnings += 1
            print(
                f"placeholder mismatch: {key}: {src_ph} -> {dst_ph}",
                file=sys.stderr,
            )
    return warnings


def write_json(path: Path, data: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8", newline="\n") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
        f.write("\n")


def remap_chunk(text: str, pua_map: dict[str, str]) -> str:
    return "".join(pua_map.get(ch, ch) for ch in text)


def remap_text(text: str, pua_map: dict[str, str]) -> str:
    out: list[str] = []
    last = 0
    for match in PLACEHOLDER.finditer(text):
        out.append(remap_chunk(text[last : match.start()], pua_map))
        out.append(match.group(0))
        last = match.end()
    out.append(remap_chunk(text[last:], pua_map))
    return "".join(out)


def remap_lang(data: dict, pua_map: dict[str, str]) -> dict:
    remapped = {}
    for key, value in data.items():
        if isinstance(value, str):
            remapped[key] = remap_text(value, pua_map)
        else:
            remapped[key] = value
    return remapped


def linzi_mcmeta_text(pua_map: dict[str, str]) -> str:
    mcmeta = load_json(PACK_MCMETA)
    mcmeta["pack"]["description"] = remap_text(LINZI_DESCRIPTION, pua_map)
    languages = mcmeta.get("language")
    if isinstance(languages, dict):
        for info in languages.values():
            if not isinstance(info, dict):
                continue
            for key in ("name", "region"):
                value = info.get(key)
                if isinstance(value, str):
                    info[key] = remap_text(value, pua_map)
    return json.dumps(mcmeta, ensure_ascii=False, indent=2) + "\n"


def linzi_font_files(font_dir: Path) -> list[tuple[Path, str]]:
    default_json = font_dir / "default.json"
    if not default_json.is_file():
        raise SystemExit(
            f"missing {default_json}; run scripts/build_linzi_font.py first"
        )
    files = [(default_json, "assets/minecraft/font/default.json")]
    pngs = sorted(font_dir.glob("linzi_e*.png"))
    if not pngs:
        raise SystemExit(f"no linzi_e*.png in {font_dir}")
    for png in pngs:
        files.append((png, f"assets/minecraft/textures/font/{png.name}"))
    return files


def build_zip(
    mcmeta_text: str,
    lang_code: str,
    lang_json: Path,
    dest: Path,
    extra: list[tuple[Path, str]] | None = None,
) -> None:
    dest.parent.mkdir(parents=True, exist_ok=True)
    lang_arcname = f"assets/minecraft/lang/{lang_code}.json"
    with zipfile.ZipFile(dest, "w", compression=zipfile.ZIP_DEFLATED) as zf:
        zf.writestr("pack.mcmeta", mcmeta_text)
        zf.write(lang_json, arcname=lang_arcname)
        for src, arcname in extra or []:
            zf.write(src, arcname=arcname)


LANG_META_KEYS = ("language.name", "language.region", "language.code")


def count_progress(source: dict, merged: dict) -> tuple[int, int]:
    compared = [key for key in source if key not in LANG_META_KEYS]
    untranslated = sum(1 for key in compared if merged.get(key) == source[key])
    translated = len(compared) - untranslated
    return translated, untranslated


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--translations",
        type=Path,
        help="Crowdin JSON (default: translations/pz_ai.json)",
    )
    parser.add_argument(
        "--linzi",
        action="store_true",
        help="remap kanji to PUA and embed the linzi bitmap font",
    )
    parser.add_argument(
        "--output",
        type=Path,
        help="zip path (default: dist/pekzep-1.16.1.zip or pekzep-linzi-1.16.1.zip)",
    )
    parser.add_argument(
        "--font-dir",
        type=Path,
        default=DIST_FONT,
        help="directory with default.json and linzi_e*.png",
    )
    args = parser.parse_args()
    if args.output is None:
        args.output = DIST_DIR / (LINZI_ZIP_NAME if args.linzi else ZIP_NAME)

    source = load_json(SOURCE_EN)
    mcmeta = load_json(PACK_MCMETA)
    lang_code, name, region = language_meta(mcmeta)
    trans_path = find_translation_file(args.translations)
    translations = load_json(trans_path)

    merged = merge(source, translations)
    merged["language.name"] = name
    merged["language.region"] = region
    merged["language.code"] = lang_code

    extra: list[tuple[Path, str]] = []
    pua_count = 0
    pua_map: dict[str, str] | None = None
    if args.linzi:
        pua_path = args.font_dir / "pua_map.json"
        if not pua_path.is_file():
            raise SystemExit(
                f"missing {pua_path}; run scripts/build_linzi_font.py first"
            )
        pua_map = load_json(pua_path)
        merged = remap_lang(merged, pua_map)
        pua_count = sum(1 for value in merged.values() if isinstance(value, str) and any("\ue000" <= ch <= "\uf8ff" for ch in value))
        extra = linzi_font_files(args.font_dir)

    placeholder_warnings = warn_placeholders(source, merged)
    translated, untranslated = count_progress(source, merged)

    staging = DIST_DIR / ("pack_linzi" if args.linzi else "pack")
    if staging.exists():
        for child in staging.rglob("*"):
            if child.is_file():
                child.unlink()
    lang_out = staging / "assets" / "minecraft" / "lang" / f"{lang_code}.json"
    write_json(lang_out, merged)

    if args.linzi:
        assert pua_map is not None
        mcmeta_text = linzi_mcmeta_text(pua_map)
    else:
        mcmeta_text = PACK_MCMETA.read_text(encoding="utf-8")
        if not mcmeta_text.endswith("\n"):
            mcmeta_text += "\n"
    build_zip(mcmeta_text, lang_code, lang_out, args.output, extra)

    stats = (
        f"keys: {len(source)}\n"
        f"translated (value differs from en_us, excl. language.*): {translated}\n"
        f"still English: {untranslated}\n"
        f"placeholder warnings: {placeholder_warnings}\n"
        f"source translations: {trans_path.relative_to(ROOT) if trans_path.is_relative_to(ROOT) else trans_path}\n"
        f"linzi: {args.linzi}\n"
        f"strings with PUA: {pua_count}\n"
        f"output: {args.output}\n"
    )
    stats_path = DIST_DIR / ("stats-linzi.txt" if args.linzi else "stats.txt")
    stats_path.write_text(stats, encoding="utf-8")
    print(stats, end="")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/build_pack.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");
const yazl = require("yazl");

const ROOT = path.resolve(__dirname, "..");
const PACK_MCMETA = path.join(ROOT, "pack", "pack.mcmeta");
const TRANSLATIONS_DIR = path.join(ROOT, "translations");
const DIST_DIR = path.join(ROOT, "dist");
const DIST_FONT = path.join(DIST_DIR, "linzi_font");
const ZIP_NAME = "pekzep-1.16.1.zip";
const LINZI_ZIP_NAME = "pekzep-linzi-1.16.1.zip";
const LINZI_DESCRIPTION = "牌言・燐字 (Pekzep linzi) for Minecraft 1.16.1";
const PLACEHOLDER = /%(?:\d+\$)?[sd]|%%/g;

function parseArgs(argv) {
  const args = {
    translations: null,
    linzi: false,
    output: null,
    fontDir: DIST_FONT,
  };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--translations") {
      args.translations = path.resolve(argv[++i]);
    } else if (a === "--linzi") {
      args.linzi = true;
    } else if (a === "--output") {
      args.output = path.resolve(argv[++i]);
    } else if (a === "--font-dir") {
      args.fontDir = path.resolve(argv[++i]);
    } else {
      throw new Error(`unknown argument: ${a}`);
    }
  }
  return args;
}

function loadJson(filePath) {
  const data = JSON.parse(fs.readFileSync(filePath, "utf8"));
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    throw new Error(`${filePath} is not a JSON object`);
  }
  return data;
}

function walkJsonFiles(dir, acc) {
  if (!fs.existsSync(dir)) {
    return acc;
  }
  for (const name of fs.readdirSync(dir)) {
    const full = path.join(dir, name);
    const st = fs.statSync(full);
    if (st.isDirectory()) {
      walkJsonFiles(full, acc);
    } else if (st.isFile() && name.endsWith(".json")) {
      acc.push(full);
    }
  }
  return acc;
}

function findTranslationFile(explicit) {
  if (explicit) {
    if (!fs.existsSync(explicit) || !fs.statSync(explicit).isFile()) {
      throw new Error(`translation file not found: ${explicit}`);
    }
    return explicit;
  }

  const preferred = [
    path.join(TRANSLATIONS_DIR, "pz_ai.json"),
    path.join(TRANSLATIONS_DIR, "pz_ai", "en_us.json"),
    path.join(TRANSLATIONS_DIR, "pz-ai", "en_us.json"),
    path.join(TRANSLATIONS_DIR, "pz-AI", "en_us.json"),
  ];
  for (const filePath of preferred) {
    if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) {
      return filePath;
    }
  }

  if (fs.existsSync(TRANSLATIONS_DIR) && fs.statSync(TRANSLATIONS_DIR).isDirectory()) {
    const candidates = walkJsonFiles(TRANSLATIONS_DIR, []).sort();
    if (candidates.length === 1) {
      return candidates[0];
    }
    if (candidates.length > 0) {
      const pz = candidates.filter((p) =>
        path.relative(TRANSLATIONS_DIR, p).toLowerCase().includes("pz")
      );
      if (pz.length === 1) {
        return pz[0];
      }
      const names = candidates.map((p) => path.relative(ROOT, p)).join(", ");
      throw new Error(
        "multiple translation JSON files; pass --translations. found: " + names
      );
    }
  }

  throw new Error("no translation file. download from Crowdin into translations/");
}

function languageMeta(mcmeta) {
  const langs = mcmeta.language;
  const keys = langs && typeof langs === "object" ? Object.keys(langs) : [];
  if (keys.length !== 1) {
    throw new Error("pack.mcmeta must define exactly one language");
  }
  const code = keys[0];
  const info = langs[code];
  if (info === null || typeof info !== "object" || Array.isArray(info)) {
    throw new Error(`invalid language entry for ${code}`);
  }
  const name = info.name;
  const region = info.region;
  if (!name || !region) {
    throw new Error("language name and region are required in pack.mcmeta");
  }
  return [code, name, region];
}

function writeJson(filePath, data) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2) + "\n", "utf8");
}

function remapChunk(text, puaMap) {
  let out = "";
  for (const ch of text) {
    out += Object.prototype.hasOwnProperty.call(puaMap, ch) ? puaMap[ch] : ch;
  }
  return out;
}

function remapText(text, puaMap) {
  const out = [];
  let last = 0;
  for (const match of text.matchAll(PLACEHOLDER)) {
    out.push(remapChunk(text.slice(last, match.index), puaMap));
    out.push(match[0]);
    last = match.index + match[0].length;
  }
  out.push(remapChunk(text.slice(last), puaMap));
  return out.join("");
}

function remapLang(data, puaMap) {
  const remapped = {};
  for (const [key, value] of Object.entries(data)) {
    remapped[key] = typeof value === "string" ? remapText(value, puaMap) : value;
  }
  return remapped;
}

function linziMcmetaText(puaMap) {
  const mcmeta = loadJson(PACK_MCMETA);
  // Description is shown before the pack is applied, so keep kanji (no PUA).
  mcmeta.pack.description = LINZI_DESCRIPTION;
  const languages = mcmeta.language;
  if (languages && typeof languages === "object") {
    for (const info of Object.values(languages)) {
      if (info === null || typeof info !== "object" || Array.isArray(info)) {
        continue;
      }
      for (const key of ["name", "region"]) {
        if (typeof info[key] === "string") {
          info[key] = remapText(info[key], puaMap);
        }
      }
    }
  }
  return JSON.stringify(mcmeta, null, 2) + "\n";
}

function linziFontFiles(fontDir) {
  const defaultJson = path.join(fontDir, "default.json");
  if (!fs.existsSync(defaultJson) || !fs.statSync(defaultJson).isFile()) {
    throw new Error(`missing ${defaultJson}; run scripts/build_linzi_font.js first`);
  }
  const files = [[defaultJson, "assets/minecraft/font/default.json"]];
  const uniformJson = path.join(fontDir, "uniform.json");
  if (!fs.existsSync(uniformJson) || !fs.statSync(uniformJson).isFile()) {
    throw new Error(`missing ${uniformJson}; run scripts/build_linzi_font.js first`);
  }
  files.push([uniformJson, "assets/minecraft/font/uniform.json"]);
  const pngs = fs
    .readdirSync(fontDir)
    .filter((name) => /^linzi_e.*\.png$/.test(name))
    .sort()
    .map((name) => path.join(fontDir, name));
  if (pngs.length === 0) {
    throw new Error(`no linzi_e*.png in ${fontDir}`);
  }
  for (const png of pngs) {
    files.push([png, `assets/minecraft/textures/font/${path.basename(png)}`]);
  }
  return files;
}

function hasPua(text) {
  for (const ch of text) {
    const cp = ch.codePointAt(0);
    if (cp >= 0xe000 && cp <= 0xf8ff) {
      return true;
    }
  }
  return false;
}

function isUnder(root, filePath) {
  const rel = path.relative(root, filePath);
  return rel !== "" && !rel.startsWith("..") && !path.isAbsolute(rel);
}

function rmFiles(dir) {
  if (!fs.existsSync(dir)) {
    return;
  }
  const walk = (current) => {
    for (const name of fs.readdirSync(current)) {
      const full = path.join(current, name);
      if (fs.statSync(full).isDirectory()) {
        walk(full);
      } else {
        fs.unlinkSync(full);
      }
    }
  };
  walk(dir);
}

function writeZip(dest, entries) {
  return new Promise((resolve, reject) => {
    const zipfile = new yazl.ZipFile();
    const out = fs.createWriteStream(dest);
    zipfile.outputStream.on("error", reject);
    out.on("error", reject);
    out.on("close", resolve);
    zipfile.outputStream.pipe(out);
    for (const { buffer, src, arcname } of entries) {
      if (buffer !== undefined) {
        zipfile.addBuffer(Buffer.from(buffer), arcname);
      } else {
        zipfile.addFile(src, arcname);
      }
    }
    zipfile.end();
  });
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  if (args.output === null) {
    args.output = path.join(DIST_DIR, args.linzi ? LINZI_ZIP_NAME : ZIP_NAME);
  }

  const mcmeta = loadJson(PACK_MCMETA);
  const [langCode, name, region] = languageMeta(mcmeta);
  const transPath = findTranslationFile(args.translations);
  let merged = loadJson(transPath);
  merged["language.name"] = name;
  merged["language.region"] = region;
  merged["language.code"] = langCode;

  let extra = [];
  let puaCount = 0;
  let puaMap = null;
  if (args.linzi) {
    const puaPath = path.join(args.fontDir, "pua_map.json");
    if (!fs.existsSync(puaPath) || !fs.statSync(puaPath).isFile()) {
      throw new Error(`missing ${puaPath}; run scripts/build_linzi_font.js first`);
    }
    puaMap = loadJson(puaPath);
    merged = remapLang(merged, puaMap);
    puaCount = Object.values(merged).filter(
      (value) => typeof value === "string" && hasPua(value)
    ).length;
    extra = linziFontFiles(args.fontDir);
  }

  const staging = path.join(DIST_DIR, args.linzi ? "pack_linzi" : "pack");
  rmFiles(staging);
  const langOut = path.join(
    staging,
    "assets",
    "minecraft",
    "lang",
    `${langCode}.json`
  );
  writeJson(langOut, merged);

  let mcmetaText;
  if (args.linzi) {
    mcmetaText = linziMcmetaText(puaMap);
  } else {
    mcmetaText = fs.readFileSync(PACK_MCMETA, "utf8");
    if (!mcmetaText.endsWith("\n")) {
      mcmetaText += "\n";
    }
  }

  fs.mkdirSync(path.dirname(args.output), { recursive: true });
  const zipEntries = [
    { buffer: mcmetaText, arcname: "pack.mcmeta" },
    { src: langOut, arcname: `assets/minecraft/lang/${langCode}.json` },
  ];
  for (const [src, arcname] of extra) {
    zipEntries.push({ src, arcname });
  }
  await writeZip(args.output, zipEntries);

  const sourceLabel = isUnder(ROOT, transPath)
    ? path.relative(ROOT, transPath)
    : transPath;
  const stats =
    `keys: ${Object.keys(merged).length}\n` +
    `source translations: ${sourceLabel}\n` +
    `linzi: ${args.linzi ? "True" : "False"}\n` +
    `strings with PUA: ${puaCount}\n` +
    `output: ${args.output}\n`;
  const statsPath = path.join(
    DIST_DIR,
    args.linzi ? "stats-linzi.txt" : "stats.txt"
  );
  fs.writeFileSync(statsPath, stats, "utf8");
  process.stdout.write(stats);
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});
